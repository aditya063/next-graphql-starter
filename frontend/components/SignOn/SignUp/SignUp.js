import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';
import * as Styled from './SignUp.style';

export const SIGN_UP_MUTATION = gql`
  mutation SIGN_UP_MUTATION($email: String!, $password: String!) {
    signUp(email: $email, password: $password) {
      id
    }
  }
`;

class SignUp extends React.PureComponent {
  state = {
    email: '',
    password: ''
  };

  onChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  onSubmit = async (e, signUp) => {
    e.preventDefault();
    await signUp();
    this.setState({ email: '', password: '' });
  };

  render() {
    const { email, password } = this.state;
    const isInvalid = email === '' || password === '';

    return (
      <Mutation mutation={SIGN_UP_MUTATION} variables={this.state}>
        {(signUp, { loading, error, data }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <p>Error: {error.message}</p>;
          // eslint-disable-next-line no-console
          if (data) console.log('data: ', data);
          return (
            <Styled.div>
              <form onSubmit={e => this.onSubmit(e, signUp)}>
                <fieldset disabled={loading} aria-busy={loading}>
                  <h2>Create a new Account</h2>

                  <label htmlFor="email">
                    Email
                    <input
                      type="email"
                      name="email"
                      placeholder="Email"
                      value={email}
                      onChange={this.onChange}
                    />
                  </label>

                  <label htmlFor="password">
                    Password
                    <input
                      type="password"
                      name="password"
                      placeholder="Password"
                      value={password}
                      onChange={this.onChange}
                    />
                  </label>

                  <button type="submit" disabled={isInvalid}>
                    Sign Up
                  </button>
                </fieldset>
              </form>
            </Styled.div>
          );
        }}
      </Mutation>
    );
  }
}

export default SignUp;
