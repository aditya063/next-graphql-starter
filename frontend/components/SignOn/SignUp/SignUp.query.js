import gql from 'graphql-tag';

import { USERS_QUERY } from '../../Users/Users.query';
import { ME_QUERY } from '../SignOn.query';

export { USERS_QUERY, ME_QUERY };

export const SIGN_UP_MUTATION = gql`
  mutation SIGN_UP_MUTATION(
    $email: String!
    $password: String!
    $confirmPassword: String!
  ) {
    signUp(
      email: $email
      password: $password
      confirmPassword: $confirmPassword
    ) {
      id
    }
  }
`;
