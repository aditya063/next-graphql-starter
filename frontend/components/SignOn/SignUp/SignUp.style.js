import { formContainer } from '../../../styles/formContainer.style';
import { greenBtn } from '../../../styles/button.style';

export { formContainer, greenBtn };
