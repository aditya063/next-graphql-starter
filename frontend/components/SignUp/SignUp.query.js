import gql from 'graphql-tag';

export const SIGN_UP_MUTATION = gql`
  mutation SIGN_UP_MUTATION(
    $email: String!
    $password: String!
    $confirmPassword: String!
  ) {
    signUp(
      email: $email
      password: $password
      confirmPassword: $confirmPassword
    ) {
      id
    }
  }
`;

export const MOCK_SIGN_UP_MUTATION = {
  request: {
    query: SIGN_UP_MUTATION,
    variables: {
      email: 'mock@email.com',
      password: 'mockpass',
      confirmPassword: 'mockpass'
    }
  },
  result: {
    data: {
      signUp: [
        {
          id: '1'
        }
      ]
    }
  }
};
