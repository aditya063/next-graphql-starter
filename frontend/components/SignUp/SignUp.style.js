import styled from 'styled-components';
import { Form as formikForm } from 'formik';
import { Form, Input, Button, Icon, List, Typography } from 'antd';

export const SignUpForm = styled(formikForm)``;

export const TypographyText = styled(Typography.Text)``;

export const FormItem = styled(Form.Item)``;

export const FormItemBtn = styled(Form.Item)`
  float: right;
`;

export const InputIcon = styled(Icon)`
  color: rgba(0, 0, 0, 0.25);
`;

export const InputEmail = styled(Input)``;

export const InputPassword = styled(Input.Password)``;

export const InputConfirmPassword = styled(Input.Password)``;

export const PassListContainer = styled.div`
  display: inline-block;
`;

export const PassList = styled(List)``;

export const PassListItem = styled(List.Item)`
  padding: 0;
`;

export const ListIcon = styled(Icon)`
  margin-right: 5px;
`;

export const SubmitBtn = styled(Button)`
  float: right;
`;
