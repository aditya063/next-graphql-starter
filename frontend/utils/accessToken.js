import fetch from 'isomorphic-unfetch';
// import cookie from 'cookie';
// import jwt from 'jsonwebtoken';

let accessToken = '';

export const setAccessToken = token => (accessToken = token);

export const getAccessToken = () => accessToken;

export const clearAccessToken = () => (accessToken = '');

export const fetchAccessToken = async refreshToken => {
  const url =
    process.env.NODE_ENV === 'development'
      ? process.env.DEV_REFRESH_URL
      : process.env.PROD_REFRESH_URL;

  const options = {
    method: 'GET',
    credentials: 'include',
    headers: { cookie: `rt=${refreshToken}` }
  };

  try {
    const response = await fetch(url, options);

    const data = await response.json();

    return data.accessToken;
  } catch (error) {
    if (process.env.NODE_ENV === 'development')
      console.error('fetchAccessToken error: ', error);
  }
};

// export const verifyAccessToken = accessToken => {
//   try {
//     jwt.verify(accessToken, process.env.ACCESS_TOKEN_SECRET);
//   } catch (error) {
//     if (process.env.NODE_ENV === 'development')
//       console.error('fetchAccessToken error: ', error);
//   }
// };

// const checkAccessToken = async cookies => {
//   const accessToken = getAccessToken();

//   if (!accessToken) return true;

//   const parsedCookies = cookie.parse(cookies);

//   if (!parsedCookies.rt) return '';

//   // const response = await fetchAccessToken(parsedCookies.rt);

//   // const data = await response.json();

//   // setAccessToken(data.accessToken);
//   // setAccessTokenExpiry(data.exp);

//   // return data.accessToken;
// };
