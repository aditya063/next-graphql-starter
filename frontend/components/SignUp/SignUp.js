import { ApolloConsumer, Mutation } from 'react-apollo';

// import { DisplayLoading, DisplayError } from '..';
import DisplayLoading from '../DisplayLoading/DisplayLoading';
import DisplayError from '../DisplayError/DisplayError';
import { SIGN_UP_MUTATION } from '../../graphql/queries';
import * as sc from './SignUp.style';

class SignUp extends React.PureComponent {
  state = { email: '', password: '', confirmPassword: '' };

  handleChangeInput = e => this.setState({ [e.target.name]: e.target.value });

  handleSubmitForm = async (e, signUp) => {
    e.preventDefault();
    this.setState({ password: '', confirmPassword: '' });
    await signUp();
  };

  handleError = error => error;

  handleCompleted = client => client.writeData({ data: { isLoggedIn: true } });

  render() {
    const { email, password, confirmPassword } = this.state;
    const isInvalid = email === '' || password === '' || confirmPassword === '';

    return (
      <ApolloConsumer>
        {client => (
          <Mutation
            mutation={SIGN_UP_MUTATION}
            variables={this.state}
            onError={this.handleError}
            onCompleted={() => this.handleCompleted(client)}
          >
            {(signUp, { loading, error }) => (
              <sc.form onSubmit={e => this.handleSubmitForm(e, signUp)}>
                <fieldset disabled={loading} aria-busy={loading}>
                  <h2>Create a new Account</h2>

                  {loading && <DisplayLoading />}

                  {error && <DisplayError error={error} />}

                  <label htmlFor="email">
                    Email
                    <sc.inputText
                      type="email"
                      name="email"
                      placeholder="email"
                      value={email}
                      onChange={this.handleChangeInput}
                    />
                  </label>

                  <label htmlFor="password">
                    Password
                    <sc.inputText
                      type="password"
                      name="password"
                      placeholder="password"
                      value={password}
                      onChange={this.handleChangeInput}
                    />
                  </label>

                  <label htmlFor="confirmPassword">
                    Confirm Password
                    <sc.inputText
                      type="password"
                      name="confirmPassword"
                      placeholder="password"
                      value={confirmPassword}
                      onChange={this.handleChangeInput}
                    />
                  </label>

                  <sc.h3PassTitle>Password must contain:</sc.h3PassTitle>

                  <sc.ulPassList aria-label="Password must contain:">
                    <li>at least 8 charactors</li>
                    <li>an uppercase letter</li>
                    <li>a lowercase letter</li>
                    <li>a number</li>
                  </sc.ulPassList>

                  <sc.inputBtn
                    type="submit"
                    value="Sign Up"
                    disabled={isInvalid}
                  />
                </fieldset>
              </sc.form>
            )}
          </Mutation>
        )}
      </ApolloConsumer>
    );
  }
}

export default SignUp;
