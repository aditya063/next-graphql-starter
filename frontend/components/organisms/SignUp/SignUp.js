import Router from 'next/router';
import { Formik, Field } from 'formik';
import { useMutation } from '@apollo/react-hooks';
import * as yup from 'yup';
import DisplayError from '../../molecules/DisplayError/DisplayError';
import { SIGN_UP } from '../../../graphql/queries';
import { setAccessToken } from '../../../utils/accessToken';
import { passwordRequirements } from '../../../constants';
import * as sc from './SignUp.style';

const validationSchema = yup.object({
  email: yup
    .string()
    .email('Invalid email')
    .max(255, 'Must be 255 characters or less')
    .required('Required'),

  password: yup
    .string()
    .min(8, 'Must be at least 8 characters')
    .max(255, 'Must be 255 characters or less')
    .required('Required'),

  confirmPassword: yup
    .string()
    .min(8, 'Must be at least 8 characters')
    .max(255, 'Must be 255 characters or less')
    .required('Required')
});

const SignUp = () => {
  const handleCompleted = data => {
    if (data && data.signUp && data.signUp.accessToken) {
      setAccessToken(data.signUp.accessToken);

      Router.push('/');
    }
  };

  const [signUp, { loading, error }] = useMutation(SIGN_UP, {
    update(cache) {
      cache.writeData({ data: { isLoggedIn: true } });
    },
    onCompleted(data) {
      handleCompleted(data);
    },
    onError(_error) {}
  });

  const handleSubmitForm = async (values, formikHelpers) => {
    signUp({ variables: values });

    formikHelpers.resetForm();
  };

  return (
    <Formik
      initialValues={{ email: '', password: '', confirmPassword: '' }}
      validationSchema={validationSchema}
      validateOnChange={false}
      validateOnBlur={true}
      onSubmit={handleSubmitForm}
    >
      {formikProps => (
        <sc.SignUpForm>
          <h2>Create a new Account</h2>

          <Field name='email'>
            {fieldProps => (
              <sc.FormItem
                label='Email'
                htmlFor='signUpEmail'
                help={fieldProps.meta.touched && fieldProps.meta.error}
                validateStatus={
                  fieldProps.meta.touched && fieldProps.meta.error
                    ? 'error'
                    : ''
                }
              >
                <sc.InputEmail
                  id='signUpEmail'
                  type='email'
                  onPressEnter={fieldProps.handleSubmit}
                  prefix={<sc.InputIcon type='user' />}
                  {...fieldProps.field}
                />
              </sc.FormItem>
            )}
          </Field>

          <Field name='password'>
            {fieldProps => (
              <sc.FormItem
                label='Password'
                htmlFor='signUpPassword'
                help={fieldProps.meta.touched && fieldProps.meta.error}
                validateStatus={
                  fieldProps.meta.touched && fieldProps.meta.error
                    ? 'error'
                    : ''
                }
              >
                <sc.InputPassword
                  id='signUpPassword'
                  onPressEnter={fieldProps.handleSubmit}
                  prefix={<sc.InputIcon type='lock' />}
                  {...fieldProps.field}
                />
              </sc.FormItem>
            )}
          </Field>

          <Field name='confirmPassword'>
            {fieldProps => (
              <sc.FormItem
                label='Confirm Password'
                htmlFor='signUpConfirmPassword'
                help={fieldProps.meta.touched && fieldProps.meta.error}
                validateStatus={
                  fieldProps.meta.touched && fieldProps.meta.error
                    ? 'error'
                    : ''
                }
              >
                <sc.InputPassword
                  id='signUpConfirmPassword'
                  onPressEnter={fieldProps.handleSubmit}
                  prefix={<sc.InputIcon type='lock' />}
                  {...fieldProps.field}
                />
              </sc.FormItem>
            )}
          </Field>

          {error && <DisplayError error={error} />}

          <sc.PassListContainer data-testid='passList'>
            <sc.TypographyText strong>
              {passwordRequirements.title}
            </sc.TypographyText>

            <sc.PassList
              split={false}
              dataSource={passwordRequirements.reqs}
              renderItem={item => (
                <sc.PassListItem>
                  <sc.ListIcon type='minus' />

                  <sc.TypographyText>{item}</sc.TypographyText>
                </sc.PassListItem>
              )}
            />
          </sc.PassListContainer>

          <sc.FormItemBtn>
            <sc.SubmitBtn
              aria-label='submit button'
              loading={loading}
              type='primary'
              htmlType='submit'
              disabled={
                !formikProps.values.email ||
                !formikProps.values.password ||
                !formikProps.values.confirmPassword ||
                formikProps.errors.email ||
                formikProps.errors.password ||
                formikProps.errors.confirmPassword ||
                formikProps.isSubmitting
              }
            >
              Sign Up
            </sc.SubmitBtn>
          </sc.FormItemBtn>
        </sc.SignUpForm>
      )}
    </Formik>
  );
};

export default SignUp;
