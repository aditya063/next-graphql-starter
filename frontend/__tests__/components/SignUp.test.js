import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import { ThemeProvider } from 'styled-components';
import { load } from '../../utils/load';
import SignUp from '../../components/SignUp/SignUp';
import { MOCK_SIGN_UP_MUTATION } from '../../components/SignUp/SignUp.query';
import theme from '../../styles/theme.style';

describe('SignUp', () => {
  let mockProps;
  let mockQueries;
  let wrapper;

  beforeEach(() => {
    jest.resetAllMocks();
    mockProps = {};
    mockQueries = [MOCK_SIGN_UP_MUTATION];
    wrapper = mount(
      <MockedProvider mocks={mockQueries} addTypename={false}>
        <ThemeProvider theme={theme}>
          <SignUp {...mockProps} />
        </ThemeProvider>
      </MockedProvider>,
      {
        disableLifecycleMethods: true
      }
    );
  });

  it('matches snapshot', () => {
    const wrapSnap = wrapper.find('formstyle__form');

    // console.log(wrapper.debug());

    expect(wrapSnap).toMatchSnapshot();
  });

  it('matches snapshot - loaded', async () => {
    const wrapSnap = wrapper.find('formstyle__form');

    await load(wrapper);

    // console.log(wrapper.debug());

    expect(wrapSnap).toMatchSnapshot();
  });
});
