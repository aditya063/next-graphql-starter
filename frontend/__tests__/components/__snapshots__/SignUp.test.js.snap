// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp renders correctly 1`] = `
<formstyle__form
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Create a new Account
    </h2>
    <label
      htmlFor="email"
    >
      Email
      <SignUpstyle__inputText
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <SignUpstyle__inputText
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </label>
    <label
      htmlFor="confirmPassword"
    >
      Confirm Password
      <SignUpstyle__inputText
        name="confirmPassword"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </label>
    <SignUpstyle__h3PassTitle>
      Password must contain:
    </SignUpstyle__h3PassTitle>
    <SignUpstyle__ulPassList
      aria-label="Password must contain:"
    >
      <li>
        at least 8 charactors
      </li>
      <li>
        an uppercase letter
      </li>
      <li>
        a lowercase letter
      </li>
      <li>
        a number
      </li>
    </SignUpstyle__ulPassList>
    <SignUpstyle__inputBtn
      disabled={true}
      type="submit"
      value="Sign Up"
    />
  </fieldset>
</formstyle__form>
`;
